### ISSUE TITLE
FM's on simple Sklearn's boston data giving NaN's

### ISSUE THREAD

**Issue by silkspace** [coreylynch/pyFM#3]
This is giving errors, am I missing something?

>>>from scipy import sparse
>>>from sklearn.datasets import load_boston
>>>import pylibfm

#instantiate FM instance with 7 latent factors
>>>fm = pylibfm.FM(num_factors=7,  num_iter=6, verbose=True)
#load dataset
>>>boston = load_boston()
#fit FM, making sure to wrap the ndarray as a sparse csr
>>>fm.fit(sparse.csr_matrix(boston.data), boston.target)

Creating validation dataset of 0.01 of training for adaptive regularization
-- Epoch 1
Training log loss: nan
-- Epoch 2
Training log loss: nan
-- Epoch 3
Training log loss: nan
-- Epoch 4
Training log loss: nan
-- Epoch 5
Training log loss: nan
-- Epoch 6
Training log loss: nan

#fm.v is also all nan.

**Comment by trunghlt** [comment]
+1

**Comment by ruifpmaia** [comment]
Did anyone actually run the Movielens example? I'm getting errors with kernel crash, maybe access violation. And it's just about using the example code and data (from 100k movielens).

**Comment by silkspace** [comment]
I am forking it and working on a patch. I'll let you know if it works.

If you normalize the boston data set it seems to work...that is strange...

Other sets seem to get alright answers, but there doesn't seem to be much
rhyme or reason to when that happens.

I am working on a few other improvements (and changing some of the
namespace).

Best, Alex

**Comment by ruifpmaia** [comment]
Hello (silkspace)
Did you manage to identify the problem?
regards
maia

**Comment by coreylynch** [comment]
Hi @silkspace, yea feature scaling is a fairly typical preprocessing step for many machine learning problems. See this for more detail: https://class.coursera.org/ml-003/lecture/21. My guess is that the unnormalized feature space blew up the gradients, but I'm going to take a closer look at this.

The following code works:
```
from scipy import sparse
from sklearn.datasets import load_boston
from sklearn.preprocessing import normalize
import pylibfm
fm = pylibfm.FM(num_factors=7, num_iter=6, verbose=True)
boston = load_boston()
fm = pylibfm.FM(num_factors=7, num_iter=6, verbose=True, initial_learning_rate=0.0001)
fm.fit(normalize(sparse.csr_matrix(boston.data)), boston.target)

Creating validation dataset of 0.01 of training for adaptive regularization
-- Epoch 1
Training log loss: 0.96985
-- Epoch 2
Training log loss: 0.96064
-- Epoch 3
Training log loss: 0.95156
-- Epoch 4
Training log loss: 0.94262
-- Epoch 5
Training log loss: 0.93383
-- Epoch 6
Training log loss: 0.92521
```
@ruifpmaia I just reran the movielens example on my laptop and wasn't able to see a problem. Would you mind opening a new issue with some steps to reproduce your error? Thanks!

**Comment by silkspace** [comment]
Hi Corey, Thanks. We rewrote the whole shebang in the ALS formulation and
now everything is fine. Not looking further into this. Thanks!

**Comment by coreylynch** [comment]
@silkspace I forgot to mention also that when trying this out on different datasets, the default settings may not be the best. We typically use cross validation to find suitable values for things like learning rate.

**Comment by silkspace** [comment]
Thanks Corey,

We did the same (grid search + cross val) to find the hypers. What is
interesting is that the ALS version, which does not learn the hypers, seems
to consistently give the lowest log loss. I found this surprising, but it
turns out that Rendle et al have already discovered this
<http://t.signauxhuit.com/e1t/c/5/f18dQhb0S7lC8dDMPbW2n0x6l2B9nMJW7t5XX48p_7DxW7fsWKY7fZkV0W3LqWkM56dwDFf5JHDY002?t=http%3A%2F%2Fwww.ismll.uni-hildesheim.de%2Fpub%2Fpdfs%2FRendle_et_al2011-Context_Aware.pdf&si=4783836553543680&pi=c9ab5608-4780-4048-d74b-0fc0ef959794>,
however, I am not sure if this is a general result for all data sets.

Best, Silk

**Comment by MLnick** [comment]
@silkspace are you going to release your fork? Or maybe PR against this repo for adding the ALS formulation?

**Comment by silkspace** [comment]
Hi All,

We started over from scratch (Matt and I). It's not the same code.

I just recently noticed this,

https://github.com/ibayer/fastFM

Best, Alex

**Comment by MLnick** [comment]
Thanks - will take a look at that link.




Still, are you gonna open-source your new version?



—

**Comment by silkspace** [comment]
Hi Corey,

Not sure as it was a work product that my company now technically owns. We
have been thinking about open sourcing our toolkit. I'll let you know.

Best, Alex

**Comment by MLnick** [comment]
Cool (I'm Nick by the way - Corey is the library author :)




The lib you linked to is pretty comprehensive, looks really good. Will test it out.

Thanks!



—

**Comment by silkspace** [comment]
Hi Nick, sorry for the mix up!

Yeah, I think the lib I linked to has more functionality than my current
implementation. Looking to build that out, client work get's in the way ;->
(a good problem to have I suppose).

Best, Alex

**Comment by sossix** [comment]
Hi @coreylynch  (or anyone else who knows how to deal with my issue), I had the same problem as silkspace (I had to deal with NaN log loss). Once the input data are normalized (as Corey said in a previous answer) it is working well. However, once I have trained my model I don't know how to predict. Indeed when I use the fm.predict it gives me responses way too small. Even on the "boston.data", ie fm.predict(normalize(sparse.csr_matrix(boston.data)) does not give me at all the boston.target.
Does anyone know what I am doing wrong ?

Thank you all a lot for your help.

Best, Solal

**Comment by tkuTM** [comment]
He guys..
i also ran into the nan issues, but with my own dataset.
The cause was that my initial_learning_rate parameter was set too high and with increasing number of train samples the parameters exploded.

**Comment by arogozhnikov** [comment]
Just in case - I've run some simple benchmarks of pylibFM vs other LibFM implementations without tuning parameters and it gives bad results (much slower, fails on large datasets, etc.)

[Post](https://arogozhnikov.github.io/2016/02/15/TestingLibFM.html) with comparison and results

Sadly, original LibFM easily won this competition.

If developers of pylibFM are interested, the code of benchmarks may be found [here](https://github.com/arogozhnikov/cats/tree/master/notebooks).
